roachtest: assign adminui ports dynamically for virtual clusters

DarrylWong opened the pull request:
This was originally removed in #115599 due to #114097 merging, but adminui was reverted in #117141 and mistakenly did not revert the special case for virtual clusters. We unskip the multitenant/distsql tests as well.

Release note: None
Fixes: #117150
Fixes: #117149
Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/117505)
<!-- Reviewable:end -->

DarrylWong commented:
[multitenant/distsql test runs](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/13454681?hideProblemsFromDependencies=false&hideTestsFromDependencies=false)

DarrylWong committed:
roachtest: assign adminui ports dynamically for virtual clusters

This was originally removed in #115599 due to #114097 merging,
but adminui was reverted in #117141 and mistakenly did not
revert the special case for virtual clusters.

Release note: None

DarrylWong committed:
roachtest: discover the adminUIPort for acceptance/gossip/restart-node-one

This test relies on assumption `adminUIPort = SQLPort + 1` but adminui is
temporarily hardcoded as 26258 while SQLPort will be dynamically assigned
without the following override. This changes it to discover the port instead
of relying on this assumption.

Release note: None

renatolabs commented:
LGTM, but probably worth waiting for #117480 to merge and then unskip those tests in this PR.

srosenberg commented on pkg/cmd/roachtest/tests/gossip.go:
```diff
@@ -338,7 +339,13 @@ func runGossipRestartNodeOne(ctx context.Context, t test.Test, c cluster.Cluster
 	settings.Env = append(settings.Env, "COCKROACH_SCAN_MAX_IDLE_TIME=5ms")
 
-	c.Start(ctx, t.L(), option.DefaultStartOpts(), settings)
+	// TODO(darrylwong): remove SetDefaultSQLPort once #117125 is addressed.
```
Given the below assumption, is the TODO still relevant? i.e., if I understood correctly, this test fails if the invariant `adminUIPort = SQLPort + 1` doesn't hold, right?

DarrylWong replied:
Yeah that's true, if we change the port assignment logic then it will fail again.

I changed it to discover the port instead, which seems more correct to me anyway.

srosenberg reviewed (approved):
Thanks!

renatolabs commented on pkg/cmd/roachtest/cluster.go:
```diff
@@ -2490,6 +2490,12 @@ func (c *clusterImpl) ExternalAdminUIAddr(
 	return c.adminUIAddr(ctx, l, node, true)
 }
 
+func (c *clusterImpl) AdminUIPort(
```
Nit: Maybe `AdminUIPorts`, since this is returning an array of ports?

renatolabs commented on pkg/cmd/roachtest/cluster.go:
```diff
@@ -2490,6 +2490,12 @@ func (c *clusterImpl) ExternalAdminUIAddr(
 }
 
+func (c *clusterImpl) AdminUIPort(
+	ctx context.Context, l *logger.Logger, node option.NodeListOption,
```
`nodes`, since that argument is a collection?

renatolabs commented on pkg/roachprod/roachprod.go:
```diff
@@ -1074,6 +1074,28 @@ func AdminURL(
 }
 
+// AdminPort finds the AdminUI ports for a cluster.
+func AdminPort(
```
Nit: plural here too.

renatolabs commented on pkg/roachprod/roachprod.go:
```diff
@@ -1074,6 +1074,28 @@ func AdminURL(
+	for _, node := range c.Nodes {
+		port, err := c.NodeUIPort(ctx, node)
+		if err != nil {
+			return nil, err
```
Nit: might be a good idea to include the `node` in the error message.

renatolabs reviewed (approved):
Thanks for the fix! Left small suggestions.

renatolabs commented on pkg/cmd/roachtest/cluster.go:
```diff
@@ -2490,6 +2490,12 @@ func (c *clusterImpl) ExternalAdminUIAddr(
+func (c *clusterImpl) AdminUIPort(
+	ctx context.Context, l *logger.Logger, node option.NodeListOption,
+) ([]int, error) {
+	return roachprod.AdminPort(ctx, l, c.MakeNodes(node), c.localCertsDir != "")
```
We can use `c.IsSecure()` here.

DarrylWong commented:
TFTRs

bors r=srosenberg, renatolabs

yuzefovich commented:
blathers backport 23.2

DarrylWong commented:
blathers backport 23.1

DarrylWong commented:
blathers backport 23.1
